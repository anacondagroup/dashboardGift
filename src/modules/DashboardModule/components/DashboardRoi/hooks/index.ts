export * from './useRoiDashboardProtector';
export * from './useRoiTable/useRoiTable';
