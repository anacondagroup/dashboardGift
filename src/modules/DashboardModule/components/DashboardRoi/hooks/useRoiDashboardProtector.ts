import { useDispatch, useSelector } from 'react-redux';
import { useEffect } from 'react';

import {
  getIsSfConnectionStateLoaded,
  getSfConnectionState,
} from '../../../../SettingsModule/store/organisation/integrations/workato/integrations/integrations.selectors';
import { ROI_ROUTES } from '../routePaths';
import { fetchSfConnectionState } from '../../../../SettingsModule/store/organisation/integrations/workato/integrations/integrations.actions';

export const useRoiDashboardProtector = (): {
  isLoaded: boolean;
  isSalesforceConnected: boolean;
  dashboard: string;
} => {
  const dispatch = useDispatch();

  const sfConnectionState = useSelector(getSfConnectionState);
  const isLoaded = useSelector(getIsSfConnectionStateLoaded);

  const dashboard = sfConnectionState === 'connected' ? ROI_ROUTES.FUNNEL : ROI_ROUTES.REPORTING;

  useEffect(() => {
    dispatch(fetchSfConnectionState());
  }, [dispatch]);

  return {
    isLoaded,
    isSalesforceConnected: sfConnectionState === 'connected',
    dashboard,
  };
};
