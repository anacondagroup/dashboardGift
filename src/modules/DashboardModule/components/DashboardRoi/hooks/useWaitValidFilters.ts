import { useEffect, useRef } from 'react';
import { TGlobalRoiFilters } from '@alycecom/services';

export const useWaitValidFilters = (filters: TGlobalRoiFilters): boolean => {
  const wasInitialRequestMadeRef = useRef(false);

  useEffect(() => {
    if (filters.teamIds.length) {
      wasInitialRequestMadeRef.current = true;
    }
  }, [filters.teamIds]);

  return wasInitialRequestMadeRef.current ? false : filters.teamIds.length === 0;
};
