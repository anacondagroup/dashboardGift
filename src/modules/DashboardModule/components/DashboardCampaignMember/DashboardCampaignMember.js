import React, { useEffect, useCallback, useMemo } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { Grid } from '@mui/material';
import { Auth, Features, User } from '@alycecom/modules';
import { TrackEvent } from '@alycecom/services';
import { useRouting, useScrollTop, useUrlQuery, useSetUrlQuery } from '@alycecom/hooks';

import DashboardHeader from '../../../../components/Dashboard/Header/DashboardHeader';
import DashboardKpi from '../Overview/DashboardKpi';
import DashboardGiftStatuses from '../Overview/DashboardGiftStatuses';
import OverviewLoader from '../../hoc/BreakdownLoaders/OverviewLoader';
import { overviewDownloadReportRequest } from '../../store/overview/overview.actions';
import DashboardSection from '../../../../components/Dashboard/Shared/DashboardSection';
import { membersLoadRequest } from '../../store/members/members.actions';
import { getMembers, getMembersIsLoading, makeGetMemberById } from '../../store/members/members.selectors';
import MemberFilter from '../Header/MemberFilter';
import { useTrackPage } from '../../../../hooks/useTrackPage';
import { getOverviewReportIsLoading } from '../../store/overview/overview.selectors';
import DashboardStickyHeader from '../Header/DashboardStickyHeader';
import { makeGetTeamIdByCampaignId } from '../../../../store/campaigns/campaigns.selectors';
import GiftInvitationReportButton from '../Breakdowns/GiftInvitationReportButton/GiftInvitationReportButton';
import GiftBreakdownSection from '../Sections/GiftBreakdownSection/GiftBreakdownSection';

const DashboardCampaignMember = ({ memberId, campaignId }) => {
  const dispatch = useDispatch();
  const go = useRouting();
  const members = useSelector(getMembers);
  const isOverviewReportLoading = useSelector(getOverviewReportIsLoading);
  const isMembersLoading = useSelector(getMembersIsLoading);
  const isUserLoading = useSelector(User.selectors.getIsUserLoading);
  const teamId = useSelector(useMemo(() => makeGetTeamIdByCampaignId(campaignId), [campaignId]));
  const { fullName: memberName = '' } = useSelector(useMemo(() => makeGetMemberById(memberId), [memberId])) || {};

  const isDashboardStatisticsEnabled = useSelector(
    Features.selectors.hasFeatureFlag(Features.FLAGS.DASHBOARD_STATISTICS_V3),
  );

  const { dateRangeFrom, dateRangeTo } = useUrlQuery(['dateRangeFrom', 'dateRangeTo']);
  const updateUrlFunc = useSetUrlQuery();
  const isHeaderLoading = isMembersLoading || isUserLoading;

  const downloadOverviewReport = useCallback(
    () => dispatch(overviewDownloadReportRequest({ memberId, teamId, campaignId, dateRangeFrom, dateRangeTo })),
    [campaignId, dateRangeFrom, dateRangeTo, dispatch, memberId, teamId],
  );

  const { trackEvent } = TrackEvent.useTrackEvent();
  const userId = useSelector(User.selectors.getUserId);
  const adminId = useSelector(Auth.selectors.getLoginAsAdminId);
  const handleChangeDateRange = useCallback(
    ({ from, to, preset }) => {
      updateUrlFunc({ dateRangeFrom: from, dateRangeTo: to });
      trackEvent(
        'Date range - Changed',
        { page: 'campaign member view', preset, userId, adminId },
        { traits: { adminId } },
      );
    },
    [adminId, trackEvent, updateUrlFunc, userId],
  );

  const handleMemberChange = useCallback(
    nextMemberId => {
      go(`/campaigns/${campaignId}/members/${nextMemberId}`);
    },
    [campaignId, go],
  );

  useScrollTop();

  useEffect(() => {
    if (teamId) {
      dispatch(membersLoadRequest(teamId));
    }
  }, [dispatch, teamId]);

  useTrackPage('Enterprise Dashboard — visit single campaign member page', {
    team_id: teamId,
    member_id: memberId,
    campaign_id: campaignId,
  });

  return (
    <>
      <OverviewLoader
        memberId={memberId}
        teamId={teamId}
        campaignId={campaignId}
        dateRangeFrom={dateRangeFrom}
        dateRangeTo={dateRangeTo}
        render={({ kpi, statuses, isLoading, total }) => (
          <>
            <DashboardStickyHeader
              kpi={kpi}
              isLoading={isLoading}
              renderFilter={() => (
                <MemberFilter
                  members={members}
                  memberId={memberId}
                  onFilterChange={handleChangeDateRange}
                  onMemberChange={handleMemberChange}
                  isLoading={isHeaderLoading || isLoading}
                  dateRangeFrom={dateRangeFrom}
                  dateRangeTo={dateRangeTo}
                />
              )}
            />
            <DashboardHeader
              header={`Let’s see how ${memberName} is doing?`}
              subHeader={`Here’s an overview of the ${memberName} performance.`}
              controls={
                <MemberFilter
                  members={members}
                  memberId={memberId}
                  onFilterChange={handleChangeDateRange}
                  onMemberChange={handleMemberChange}
                  isLoading={isHeaderLoading || isLoading}
                  dateRangeFrom={dateRangeFrom}
                  dateRangeTo={dateRangeTo}
                />
              }
            />
            <Grid container justifyContent="flex-end">
              <GiftInvitationReportButton
                campaignId={campaignId}
                memberId={memberId}
                dateRangeFrom={dateRangeFrom}
                dateRangeTo={dateRangeTo}
              />
            </Grid>
            <DashboardKpi kpi={kpi} isLoading={isLoading} />
            {!isDashboardStatisticsEnabled && (
              <DashboardSection
                hidePaper
                isLoading={isLoading}
                title={`Status of ${total} gift invites`}
                icon="heart-rate"
                subtitle={`This shows you the current status of all the gifts that have been sent in the ${memberName} campaign.`}
                isReportLoading={isOverviewReportLoading}
                onDownloadReport={downloadOverviewReport}
              >
                <DashboardGiftStatuses isLoading={isLoading} statuses={statuses} />
              </DashboardSection>
            )}
          </>
        )}
      />

      <GiftBreakdownSection
        teamId={teamId}
        memberId={memberId}
        campaignId={campaignId}
        subtitle={`This is a list of all the gifts that have been sent in the ${memberName} campaign.`}
      />
    </>
  );
};

DashboardCampaignMember.propTypes = {
  campaignId: PropTypes.number.isRequired,
  memberId: PropTypes.number.isRequired,
};

export default DashboardCampaignMember;
