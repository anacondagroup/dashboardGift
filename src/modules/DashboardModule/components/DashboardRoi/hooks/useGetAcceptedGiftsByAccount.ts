import { TAcceptedGiftByEmailDomain, useGetAcceptedGiftsByAccountQuery } from '@alycecom/services';

import { transformRoiFilters } from '../utils/transformRoiFilters';

import { TDebouncedFilters } from './useRoiTable/useRoiTable';
import { useWaitValidFilters } from './useWaitValidFilters';

export type TUseGetAcceptedGiftsByAccountParams = {
  accountId: string;
  filters: TDebouncedFilters<TAcceptedGiftByEmailDomain>;
};

export const useGetAcceptedGiftsByAccount = ({
  accountId,
  filters,
}: TUseGetAcceptedGiftsByAccountParams): ReturnType<typeof useGetAcceptedGiftsByAccountQuery> & {
  isWaitingForFilters: boolean;
} => {
  const isWaitingForFilters = useWaitValidFilters(filters);
  const queryResult = useGetAcceptedGiftsByAccountQuery(
    {
      filters: transformRoiFilters(filters),
      accountId,
    },
    { skip: isWaitingForFilters },
  );

  return {
    ...queryResult,
    isWaitingForFilters,
  };
};
