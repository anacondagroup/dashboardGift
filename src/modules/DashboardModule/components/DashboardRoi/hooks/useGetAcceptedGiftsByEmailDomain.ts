import { TAcceptedGiftByEmailDomain, useGetAcceptedGiftsByEmailDomainQuery } from '@alycecom/services';

import { transformRoiFilters } from '../utils/transformRoiFilters';

import { TDebouncedFilters } from './useRoiTable/useRoiTable';
import { useWaitValidFilters } from './useWaitValidFilters';

export type TUseGetAcceptedGiftsByEmailDomainParams = {
  emailDomain: string;
  filters: TDebouncedFilters<TAcceptedGiftByEmailDomain>;
};

export const useGetAcceptedGiftsByEmailDomain = ({
  emailDomain,
  filters,
}: TUseGetAcceptedGiftsByEmailDomainParams): ReturnType<typeof useGetAcceptedGiftsByEmailDomainQuery> & {
  isWaitingForFilters: boolean;
} => {
  const isWaitingForFilters = useWaitValidFilters(filters);
  const queryResult = useGetAcceptedGiftsByEmailDomainQuery(
    {
      filters: transformRoiFilters(filters),
      emailDomain,
    },
    { skip: isWaitingForFilters },
  );

  return {
    ...queryResult,
    isWaitingForFilters,
  };
};
