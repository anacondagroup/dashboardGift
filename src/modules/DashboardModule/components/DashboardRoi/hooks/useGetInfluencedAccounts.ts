import { TInfluencedAccounts, useGetInfluencedAccountsQuery } from '@alycecom/services';

import { transformRoiFilters } from '../utils/transformRoiFilters';

import { TDebouncedFilters } from './useRoiTable/useRoiTable';
import { useWaitValidFilters } from './useWaitValidFilters';

export const useGetInfluencedAccounts = (
  filters: TDebouncedFilters<TInfluencedAccounts>,
): ReturnType<typeof useGetInfluencedAccountsQuery> & {
  isWaitingForFilters: boolean;
} => {
  const isWaitingForFilters = useWaitValidFilters(filters);
  const queryResult = useGetInfluencedAccountsQuery(transformRoiFilters(filters), { skip: isWaitingForFilters });

  return {
    ...queryResult,
    isWaitingForFilters,
  };
};
