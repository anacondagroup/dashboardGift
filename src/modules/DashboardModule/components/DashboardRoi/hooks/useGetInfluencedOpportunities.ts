import { TInfluencedOpportunities, useGetInfluencedOpportunitiesQuery } from '@alycecom/services';

import { transformRoiFilters } from '../utils/transformRoiFilters';

import { TDebouncedFilters } from './useRoiTable/useRoiTable';
import { useWaitValidFilters } from './useWaitValidFilters';

export const useGetInfluencedOpportunities = (
  filters: TDebouncedFilters<TInfluencedOpportunities>,
): ReturnType<typeof useGetInfluencedOpportunitiesQuery> & {
  isWaitingForFilters: boolean;
} => {
  const isWaitingForFilters = useWaitValidFilters(filters);
  const queryResult = useGetInfluencedOpportunitiesQuery(transformRoiFilters(filters), { skip: isWaitingForFilters });

  return {
    ...queryResult,
    isWaitingForFilters,
  };
};
