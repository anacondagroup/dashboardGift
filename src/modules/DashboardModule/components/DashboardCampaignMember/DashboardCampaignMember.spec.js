import React from 'react';
import { useUrlQuery } from '@alycecom/hooks';

import { render, cleanup } from '../../../../testUtils';

import DashboardCampaignMember from './DashboardCampaignMember';

jest.mock('../Breakdowns/GiftInvitationReportButton/GiftInvitationReportButton', () => () =>
  'GiftInvitationReportButton',
);
jest.mock('../../../../components/Dashboard/Header/DashboardHeader', () => ({ controls, ...props }) => (
  <dashboard-header {...props}>
    <div>DashboardHeader</div>
    {controls}
  </dashboard-header>
));
jest.mock('../Overview/DashboardKpi', () => () => 'DashboardKPI');
jest.mock('../Overview/DashboardGiftStatuses', () => () => 'DashboardGiftStatuses');
jest.mock('../../hoc/BreakdownLoaders/OverviewLoader', () => ({ render: renderContent }) => (
  <div>{renderContent({ kpi: '#kpi#', statuses: '#statuses#', isLoading: false, total: '#total' })}</div>
));
jest.mock('../../../../components/Dashboard/Shared/DashboardSection', () => ({ children, ...props }) => (
  <dashboard-section {...props}>{children}</dashboard-section>
));
jest.mock('../Header/MemberFilter', () => () => 'MemberFilter');
jest.mock('../Header/DashboardStickyHeader', () => ({ renderFilter, ...props }) => (
  <dashboard-sticky-header {...props}>{renderFilter()}</dashboard-sticky-header>
));
jest.mock('../Sections/GiftBreakdownSection/GiftBreakdownSection', () => () => 'GiftBreakdownSection');
jest.mock('@alycecom/hooks', () => ({
  useUrlQuery: jest.fn(),
  useSetUrlQuery: jest.fn(),
  useScrollTop: jest.fn(),
  useRouting: jest.fn(),
}));

describe('DashboardCampaignMember', () => {
  const member = { id: 10 };
  const campaign = { id: 11 };
  const defaultInitialState = {
    dashboard: {
      overview: {
        isReportLoading: false,
      },
      members: {
        members: [member],
        isLoading: false,
        error: null,
      },
    },
  };

  beforeEach(() => {
    useUrlQuery.mockReset().mockReturnValue({
      contactId: 'CONTACT_ID',
      dateRangeFrom: 'DATE_RANGE_FROM',
      dateRangeTo: 'DATE_RANGE_TO',
      giftSort: 'GIFT_SORT',
      giftDirection: 'asc',
      giftSearch: 'GIFT_SEARCH',
      giftPage: 'GIFT_PAGE',
    });
  });

  afterEach(cleanup);

  const setup = (props, initialState = defaultInitialState) =>
    render(<DashboardCampaignMember campaignId={campaign.id} memberId={member.id} />, {
      initialState,
    });

  it('Should render correctly when member loaded', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render correctly while members loading', () => {
    const state = {
      ...defaultInitialState,
      dashboard: {
        ...defaultInitialState.dashboard,
        members: {
          ...defaultInitialState.dashboard.members,
          members: [],
          isLoading: true,
        },
      },
    };
    const { asFragment } = setup(undefined, state);
    expect(asFragment()).toMatchSnapshot();
  });
});
