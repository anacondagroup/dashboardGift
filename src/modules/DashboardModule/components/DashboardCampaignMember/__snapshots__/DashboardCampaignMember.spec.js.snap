// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardCampaignMember Should render correctly when member loaded 1`] = `
<DocumentFragment>
  <div>
    <dashboard-sticky-header
      isloading="false"
      kpi="#kpi#"
    >
      MemberFilter
    </dashboard-sticky-header>
    <dashboard-header
      header="Let’s see how  is doing?"
      subheader="Here’s an overview of the  performance."
    >
      <div>
        DashboardHeader
      </div>
      MemberFilter
    </dashboard-header>
    <div
      class="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
    >
      GiftInvitationReportButton
    </div>
    DashboardKPI
    <dashboard-section
      hidepaper="true"
      icon="heart-rate"
      isloading="false"
      isreportloading="false"
      subtitle="This shows you the current status of all the gifts that have been sent in the  campaign."
      title="Status of #total gift invites"
    >
      DashboardGiftStatuses
    </dashboard-section>
  </div>
  GiftBreakdownSection
</DocumentFragment>
`;

exports[`DashboardCampaignMember Should render correctly while members loading 1`] = `
<DocumentFragment>
  <div>
    <dashboard-sticky-header
      isloading="false"
      kpi="#kpi#"
    >
      MemberFilter
    </dashboard-sticky-header>
    <dashboard-header
      header="Let’s see how  is doing?"
      subheader="Here’s an overview of the  performance."
    >
      <div>
        DashboardHeader
      </div>
      MemberFilter
    </dashboard-header>
    <div
      class="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
    >
      GiftInvitationReportButton
    </div>
    DashboardKPI
    <dashboard-section
      hidepaper="true"
      icon="heart-rate"
      isloading="false"
      isreportloading="false"
      subtitle="This shows you the current status of all the gifts that have been sent in the  campaign."
      title="Status of #total gift invites"
    >
      DashboardGiftStatuses
    </dashboard-section>
  </div>
  GiftBreakdownSection
</DocumentFragment>
`;
